src/rrd/mod.rs

```diff
@@ -0,0 +1,529 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use core::time;
+use sqlx::{self, Sqlite};
+use std::{io::Write, path::Path};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("invalid window, can't be zero")]
+    InvalidWindow,
+
+    #[error("invalid retention, can't be zero or less than window")]
+    InvalidRetention,
+}
+
+/// Slot provides the functionality to set or overwrite the value of any metric
+/// at a specific timestamp.
+#[async_trait]
+pub trait Slot {
+    /// Counter sets (or overrides) the current stored value for this key,
+    /// with value
+    async fn counter(&mut self, key: &str, value: f64) -> Result<()>;
+    /// Key return the key of the slot which is the window timestamp
+    async fn key(&self) -> Result<i64>;
+}
+
+/// RRD is a round robin database of fixed size which is specified on creation
+/// RRD stores `counter` values. Which means values that can only go UP.
+/// then it's easy to compute the increase of this counter over a given window
+/// The database only keep history based on retention.
+#[async_trait]
+pub trait RRD<S, I, 'a>
+where
+    S: Slot,
```

Comment:
I really think `S: Slot` here is better done as an associated type. since for each implementation there should be only one type of Slot. 

If you have a question please let me know. we can have another quick zoom